feat: 优化 query_string 语法 --story=126609837

Zhaoyikaiii opened the pull request.

shamcleren committed:
feat: 优化 query_string 语法 #1010158081126609837

shamcleren committed:
feat: 优化 query_string 语法 #1010158081126609837

shamcleren committed:
feat: 优化 query_string 语法 #1010158081126609837

shamcleren committed:
feat: 优化 query_string 语法 #1010158081126609837

Zhaoyikaiii committed:
chore: add lucene to sql and es dsl test cases

Zhaoyikaiii committed:
feat: impl lucene to sql

Zhaoyikaiii committed:
feat(test): Enhance Elasticsearch query test utilities

Zhaoyikaiii committed:
chore: r ename ToSQl to ToSQlString

Zhaoyikaiii committed:
feat: Add Lucene to SQL parser

Zhaoyikaiii committed:
refactor: Rename QueryVisitor to Statement & enhance es related parser tests

Zhaoyikaiii committed:
refactor: Rename QueryVisitor to StatementVisitor and update related tests

Zhaoyikaiii committed:
refactor(parser): Refactor query parsing and visitor logic

shamcleren commented on pkg/unify-query/internal/lucene_parser/generate.go:
```diff
@@ -0,0 +1,28 @@
+//go:generate antlr4 -Dlanguage=Go -no-listener -visitor -package gen *.g4 -o ../gen
+//go:generate antlr4 -Dlanguage=Go -listener -no-visitor -package gen *.g4 -o ../gen
+
+// for lexer.g4
```
这一段一直往下，没有作用

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+	next = s
+
+	switch ctx.(type) {
+	case *gen.TopLevelQueryContext:
```
138 和 139 可以直接删除，跟下面逻辑重复了

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+
+	switch ctx.(type) {
+	case *gen.FieldRangeExprContext:
+		rangeNode := NewRangeNode()
```
写法跟其他的保持，使用 &RangeNode{} 的方式

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+	case *gen.TermContext:
+		// Check if this term context contains a range expression
+		termText := ctx.GetText()
+		if strings.Contains(termText, "TO") && (strings.Contains(termText, "[") || strings.Contains(termText, "{")) {
```
这个判断逻辑有点 HardCode，可以在 TermNode 里面增加一个 TermRangeExprContext 判定来解决

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+
+func (n *RangeNode) VisitChildren(ctx antlr.RuleNode) interface{} {
+	switch c := ctx.(type) {
+	case *gen.TermRangeExprContext:
```
这里一直到 1238 行，都可以删掉，代码内容重复了

Zhaoyikaiii replied:
已删除

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+
+type TermNode struct {
+	baseNode
+	field      string
```
建议所有的类型都使用 Node 接口的实现方式，即使是 String

Zhaoyikaiii replied:
已添加

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+	return
+}
+
+func (n *DisjNode) analyzeBooleanClauses2sql() (mustClauses, mustNotClauses, shouldClauses []querystring_parser.Expr) {
```
这个方法和下面 analyzeBooleanQueries 的方法可以直接在 ToEs 和 ToSQL 里面就直接解决了。

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+	return result
+}
+
+func (n *ConjNode) ToES() elastic.Query {
```
这里的命名建议使用 ToEsDsl

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+	switch ctx.(type) {
+	case *gen.ConjQueryContext:
+		conjNode := &ConjNode{}
+		n.children = append(n.children, conjNode)
```
DisjNode 和 ConjNode 的实现感觉差不多，是否可以使用 ConjNode 代替 DisjNode ？另外，命名的方式可以使用更优雅一点，例如 LogicNode

Zhaoyikaiii replied:
已修改,先处理成一个中间的Expr结构

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,1792 @@
+
+type DisjNode struct {
+	baseNode
+	children []Node
```
这里的属性命名可以使用 nodes

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	baseNode
+
+	root    Node
+	errNode []string
```
建议修改为 error 类型

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+type Statement struct {
+	baseNode
+
+	root    Node
```
建议直接使用 node 名称

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+}
+
+func (s *Statement) trans() Expr {
+	if s.root != nil {
```
s.root == nil 使用提前退出方式

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+
+type QueryNode struct {
+	baseNode
+	children []Node
```
这里可以使用 nodes 来命名

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	return result
+}
+
+type DisjNode struct {
```
这个是不是改成 OrNode 更好理解一点

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	return visitChildren(next, ctx)
+}
+
+func (n *QueryNode) buildBooleanExpression(mustClauses, mustNotClauses, shouldClauses []Expr) Expr {
```
这个实现可以抽象成一个通用的实现，然后所有需要拼接 must / should /  must_not 等等，都可以用该方法实现，例如：func (n *DisjNode) Expr() Expr { 的实现

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	return visitChildren(next, ctx)
+}
+
+func (n *ConjNode) splitClause() (mustClauses, mustNotClauses, shouldClauses []Expr) {
```
建议这块抽象成一个独立判定逻辑，从 DisjNode 层级可以直接使用即可，不用利用 ConjNode 再透传一轮

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	case *gen.FieldNameContext:
+		// Store the field name for this node
+		n.field = ctx.GetText()
+		return visitChildren(next, ctx)
```
这一段没作用

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	return expr
+}
+
+func (n *ClauseNode) VisitChildren(ctx antlr.RuleNode) interface{} {
```
参考：
func (n *ClauseNode) VisitChildren(ctx antlr.RuleNode) interface{} {
	var next Node
	next = n

	switch ctx.(type) {
	case *gen.FieldRangeExprContext:
		n.node = &RangeNode{}
		next = n.node
	case *gen.FieldNameContext:
		// Store the field name for this node
		n.field = ctx.GetText()
	case *gen.TermContext:
		// Check if this term context contains a range expression
		termText := ctx.GetText()
		if strings.Contains(termText, "TO") {
			// This is a range expression, create a RangeNode instead
			n.node = &RangeNode{
				field: n.field,
			}
		} else {
			n.node = &TermNode{}
		}
		next = n.node
	case *gen.GroupingExprContext:
		n.node = &GroupNode{}
		next = n.node
	}

	return visitChildren(next, ctx)
}

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+}
+
+// FieldTermNode handles field:node expressions
+type FieldTermNode struct {
```
这个都没地方用

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	isWildcard bool
+}
+
+func (n *TermNode) trans() Expr {
```
trans 建议改名同意成 Expr 即可

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	} else if n.isWildcard {
+		expr = &WildcardExpr{Value: &StringExpr{Value: n.value}}
+	} else {
+		matchExpr := &MatchExpr{Value: &StringExpr{Value: n.value}}
```
这一大段完全可以改成 

expr = &MatchExpr{
			Value: &StringExpr{
				Value: n.value,
			},
			IsQuoted: n.isQuoted,
		}

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+			cleanValue = unescapeString(text)
+		}
+
+		if strings.Contains(cleanValue, "*") || strings.Contains(cleanValue, "?") {
```
验证一下如果有反斜杠的情况下是否可以保留

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	return visitChildren(next, ctx)
+}
+
+func (n *TermNode) VisitTerminal(ctx antlr.TerminalNode) interface{} {
```
这里对于类型的判断跟VisitChildren 里面都有，是不是有一个是完全没用的？

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,992 @@
+	}
+
+	// Handle terminals by text for backwards compatibility
+	switch text {
```
跟上面的判断冲突了

Zhaoyikaiii committed:
refactor(parser): Rename ToES to es and update related parsing logic

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,893 @@
+
+	switch ctx.(type) {
+	case *gen.DisjQueryContext:
+		disjNode := &OrNode{}
```
改成 node，或者直接使用 next 也可以

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,893 @@
+	return visitChildren(next, ctx)
+}
+
+func buildBooleanExpression(mustClauses, mustNotClauses, shouldClauses []Expr) Expr {
```
可以给这个方法加一个单侧

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,893 @@
+
+	switch ctx.(type) {
+	case *gen.ConjQueryContext:
+		conjNode := &AndNode{}
```
直接用 node 表示即可

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/visitor.go:
```diff
@@ -0,0 +1,893 @@
+
+	switch ctx.(type) {
+	case *gen.ModClauseContext:
+		modClauseNode := &ModClauseNode{}
```
直接使用 next

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/sql.go:
```diff
@@ -0,0 +1,222 @@
+
+	case *TimeRangeExpr:
+		field := getFieldName(e.Field)
+		if field == fmt.Sprintf("`%s`", DefaultLogField) {
```
这段代码有问题

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/internal/lucene_parser/expr.go:
```diff
@@ -0,0 +1,108 @@
+	nr.Field = &StringExpr{Value: field}
+}
+
+type TimeRangeExpr struct {
```
TimeRangeExpr 这个可以和 NumberRangeExpr 以及 WildcardExpr  RegexpExpr MatchExpr 应该都可以合并成一个 OperatorExpr

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/expr.go:
```diff
@@ -0,0 +1,108 @@
+	tr.Field = &StringExpr{Value: field}
+}
+
+type ConditionExpr struct {
```
使用 ConditionsExpr

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/expr.go:
```diff
@@ -0,0 +1,108 @@
+	Values [][]Expr
+}
+
+type ConditionMatchExpr struct {
```
这个可以使用 ConditionExpr 包含 key、value、op（操作符）三者，代替上面所有的

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/sql.go:
```diff
@@ -0,0 +1,222 @@
+		return fmt.Sprintf("%v", e.Value)
+
+	case *NumberExpr:
+		return strconv.FormatFloat(e.Value, 'f', -1, 64)
```
cast.ToFloat 使用这个

shamcleren commented on pkg/unify-query/internal/lucene_parser/es.go:
```diff
@@ -0,0 +1,449 @@
+			}
+			return elastic.NewMatchQuery(field, value)
+		case FieldTypeLong, FieldTypeInteger:
+			if num, err := strconv.ParseInt(value, 10, 64); err == nil {
```
使用 cast

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/parser_test.go:
```diff
@@ -0,0 +1,718 @@
+					Value: &StringExpr{Value: "bro*"},
+				},
+			},
+			es:  `{"bool":{"should":[{"query_string":{"query":"qu?ck"}},{"query_string":{"query":"bro*"}}]}}`,
```
这里的 query_string 少了东西，{"query_string":{"analyze_wildcard":true,"fields":["*","__*"],"lenient":true,"query":"erro?"}}

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/internal/lucene_parser/es.go:
```diff
@@ -0,0 +1,449 @@
+	field := getESFieldName(e.Field)
+	value := getESValue(e.Value)
+
+	if field == DefaultLogField {
```
这里不能这样判断，es 无需补充 DefaultLogField

Zhaoyikaiii replied:
修改为:DefaultEmptyField

shamcleren commented on pkg/unify-query/internal/lucene_parser/es.go:
```diff
@@ -0,0 +1,449 @@
+			return s.Value
+		}
+	}
+	return DefaultLogField
```
这里不需要不从默认字段，只有 doris 才有该场景

Zhaoyikaiii replied:
修改为:DefaultEmptyField

Zhaoyikaiii committed:
refactor(expr): Unify query expression types with OperatorExpr

Zhaoyikaiii committed:
refactor: operatorExpr(wip)

Zhaoyikaiii committed:
fix(lucene_parser): handle empty-field wildcard/regexp queries and improve SQL formatting

- Generate QueryStringQuery when field is DefaultEmptyField for wildcard/regexp
- Quote numeric values in SQL only when they are strings
- Fix range boundary inclusion flags for > and < operators
- Update test expectations accordingly

Zhaoyikaiii committed:
refactor(lucene): enhance default-field queries with wildcard analysis

Replace direct elastic.NewQueryStringQuery calls with createEnhancedQueryStringQuery
to enable wildcard analysis, lenient parsing, and multi-field search on both the
default empty field and "__". Rename ConditionExpr to ConditionsExpr for clarity
and update test expectations accordingly.

Zhaoyikaiii committed:
test(lucene_parser): add unit tests for buildBooleanExpression

Zhaoyikaiii committed:
feat: lucenen_parser 替换querystring_parser

Zhaoyikaiii committed:
feat: enhance query handling with nested fields and field aliases

Zhaoyikaiii committed:
fix(docs): correct description field in API documentation

Zhaoyikaiii committed:
feat: optimize query string parser by refactoring schema handling and improving field type management

Zhaoyikaiii committed:
feat: optimize query string parser by improving logical operator handling and enhancing test cases

Zhaoyikaiii committed:
feat: enhance query string parser with support for colon operator and additional test cases

Zhaoyikaiii committed:
feat: 替换queryString(wip)

- 目前还不支持针对nested
- 修改失败的测试
  - keyword使用term,text使用match_phrase
  - 修改queryString为:"*", "__*"

Zhaoyikaiii committed:
chore: 修改测试用例(wip)

Zhaoyikaiii committed:
feat: optimize query string parser by enhancing nested field handling and refining field name resolution

Zhaoyikaiii committed:
feat: enhance query string parser with improved error handling and support for fuzzy matching

Zhaoyikaiii committed:
feat: optimize query string parser by improving value handling and enhancing test expectations

Zhaoyikaiii committed:
feat: 优化 query_string 语法 #1010158081126609837

Zhaoyikaiii committed:
feat: enhance query string parser with improved field option handling and initialization

Zhaoyikaiii committed:
feat: 优化 query_string 语法 #1010158081126609837

Zhaoyikaiii committed:
Merge branch 'master' into feat/es_query_string_parser_optimiztion/#1010158081126609837

Zhaoyikaiii committed:
feat: 优化 query_string 语法 #1010158081126609837

Zhaoyikaiii committed:
feat: 优化 query_string 语法 #1010158081126609837

shamcleren committed:
Merge branch 'master' into feat/es_query_string_parser_optimiztion/#1010158081126609837
